### PR TITLE
Basic Sidesheet Components

### LINKED ISSUE #46639: Create Shared Sidesheet components
Create components, storybook stories and unit tests for:

- [ ] `Sidesheet`
- [ ] `SidesheetHeader`
- [ ] `SidesheetBody`
- [ ] `SidesheetTabs`
- [ ] `SidesheetCard`
- [ ] `SidesheetCardSection`

Also need to handle "sublevels" or "pages" within a sidesheet.

All of these will be separately exported and used as the building blocks for particular Sidesheets

Screen Shot 2024-08-08 at 8 09 18 AM

### PR DESCRIPTION
**Pull request by iethree**
Closes https://github.com/metabase/metabase/issues/46639

### Description

Creates basic sidesheet components, and storybook stories for all of it

- `Sidesheet` - main container
- `SidesheetSubPage` - variant of main container with a special header
- `SidesheetCard` - card for use in sidesheets
- `SidesheetCardSection` - sections within cards
- `SidesheetButton`- idiomatic button for use in the sidesheet
- `SidesheetButtonWithChevron` - another button with a chevron that shows up a lot in designs
- `SidesheetTabPanelContainer` - so you don't have to reinvent the flexbox wheel to get tabs to stick to the top

withtabs

subpageSidesheets

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

### CODE REVIEW

File: frontend/src/metabase/common/components/Sidesheet/Sidesheet.tsx [hunk 1]
```diff
@@ -0,0 +1,79 @@
+import type React from "react";
+import { t } from "ttag";
+
+import { Modal, Stack } from "metabase/ui";
+
+import Styles from "./sidesheet.module.css";
+
+type Size = "xs" | "sm" | "md" | "lg" | "xl" | "auto";
```
Comment (rafpaf) [inline frontend/src/metabase/common/components/Sidesheet/Sidesheet.tsx:8]: Does this already exist elsewhere in the codebase?

File: frontend/src/metabase/common/components/Sidesheet/Sidesheet.tsx [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+export function Sidesheet({
+  title,
+  isOpen,
+  onClose,
+  size = "sm",
+  children,
+  removeBodyPadding,
+}: SidesheetProps) {
+  return (
+    <Modal.Root opened={!!isOpen} onClose={onClose} h="100dvh">
```
Comment (rafpaf) [inline frontend/src/metabase/common/components/Sidesheet/Sidesheet.tsx:38]: Can remove the "!!" I think

File: frontend/src/metabase/common/components/Sidesheet/Sidesheet.tsx [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+      <Modal.Overlay aria-label={t`modal overlay`} />
```
Comment (rafpaf) [inline frontend/src/metabase/common/components/Sidesheet/Sidesheet.tsx:39]: Not sure overlays need aria labels - a screen reader would never announce them

Maybe a testid could work here. It's one less string in poeditor

File: frontend/src/metabase/common/components/Sidesheet/Sidesheet.tsx [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+      <Modal.Content
+        transitionProps={{ transition: "slide-left" }}
+        px="none"
+        w={sizes[size]}
+        bg="bg-light"
```
Comment (rafpaf) [inline frontend/src/metabase/common/components/Sidesheet/Sidesheet.tsx:44]: Is this still supported?

File: frontend/src/metabase/common/components/Sidesheet/Sidesheet.tsx [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        className={Styles.SidesheetContent}
+      >
+        <Modal.Header bg="bg-light" px="xl">
```
Comment (rafpaf) [inline frontend/src/metabase/common/components/Sidesheet/Sidesheet.tsx:47]: I'm not sure - do we need to use a css color var here?

File: frontend/src/metabase/common/components/Sidesheet/Sidesheet.tsx [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+          {title && (
+            <Modal.Title py="md" pr="sm">
```
Comment (rafpaf) [inline frontend/src/metabase/common/components/Sidesheet/Sidesheet.tsx:49]: Flexbox spacing might be a bit better than pr

File: frontend/src/metabase/common/components/Sidesheet/Sidesheet.tsx [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+              {title}
+            </Modal.Title>
+          )}
+          <Modal.CloseButton aria-label={t`close`} />
```
Comment (rafpaf) [inline frontend/src/metabase/common/components/Sidesheet/Sidesheet.tsx:53]: Should we capitalize "close" - would that  match an existing string?

File: frontend/src/metabase/common/components/Sidesheet/Sidesheet.tsx [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+        </Modal.Header>
+        <Modal.Body
+          data-testid="sidesheet"
+          p={0}
+          style={{
```
Comment (rafpaf) [inline frontend/src/metabase/common/components/Sidesheet/Sidesheet.tsx:58]: CSS here maybe?

File: frontend/src/metabase/common/components/Sidesheet/Sidesheet.unit.spec.tsx [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+  it("should fire onClose when modal backdrop is clicked", async () => {
+    const closeSpy = jest.fn();
+
+    render(
+      <Sidesheet title="My Title" isOpen onClose={closeSpy}>
+        hello world
+      </Sidesheet>,
+    );
+
+    const backdrop = screen.getByLabelText("modal overlay");
```
Comment (rafpaf) [inline frontend/src/metabase/common/components/Sidesheet/Sidesheet.unit.spec.tsx:71]: Should pressing escape also close the modal?

File: frontend/src/metabase/common/components/Sidesheet/sidesheet.module.css [hunk 10]
```diff
@@ -0,0 +1,24 @@
+.SidesheetContent {
+  &:not([data-css-specificity-hack="🤷‍♀️"]) {
```
Comment (rafpaf) [inline frontend/src/metabase/common/components/Sidesheet/sidesheet.module.css:2]: Maybe add a comment explaining why this is needed

**Review by rafpaf — COMMENTED** [review]
This is exciting. Some initial thoughts:

File: frontend/src/metabase/common/components/Sidesheet/Sidesheet.tsx (hunk 1 above)
Comment (iethree) [reply frontend/src/metabase/common/components/Sidesheet/Sidesheet.tsx:8]: I don't really want to create a cross dependency between sidesheet sizes and sizes for other things. maybe we suddenly want 2xl text, but 2xl will be meaningless for sidesheets and create type errors

Comment (rafpaf) [reply frontend/src/metabase/common/components/Sidesheet/Sidesheet.tsx:8]: I concur

File: frontend/src/metabase/common/components/Sidesheet/Sidesheet.unit.spec.tsx (hunk 9 above)
Comment (rafpaf) [reply frontend/src/metabase/common/components/Sidesheet/Sidesheet.unit.spec.tsx:71]: I think this functionality already exists for modals and is tested so no need to add it here

File: frontend/src/metabase/common/components/Sidesheet/SidesheetCard.tsx [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+  styleProps?: Partial<MantineStyleSystemProps>;
+}
+
+export const SidesheetCard = ({
+  title,
+  children,
+  ...styleProps
+}: SidesheetCardProps) => {
+  return (
+    <Paper p="lg" withBorder shadow="none" {...styleProps}>
```
Comment (rafpaf) [inline frontend/src/metabase/common/components/Sidesheet/SidesheetCard.tsx:18]:
```suggestion
type SidesheetCardProps = {
  title?: string;
  children: React.ReactNode;
} & PaperProps

export const SidesheetCard = ({
  title,
  children,
  ...paperProps
}: SidesheetCardProps) => {
  return (
    <Paper p="lg" withBorder shadow="none" {...paperProps}>
```

**Review by rafpaf — APPROVED** [review]
I left a few comments previously addressing little things like `pr=` - if you haven't seen those, please have a gander - but looks good enough to me for approval.

**Pull request merged**